File: Circular_Linked_list/CircularSinglyLinkedListTraversal.java
// File Name: CircularSinglyLinkedListTraversal.java
// Program to traverse a circular singly linked list
// left to right and right to left (using recursion)

class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class CircularSinglyLinkedListTraversal {
    static Node head = null;

    // Left to Right traversal
    static void traverseForward() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }

        Node temp = head;
        System.out.print("Left to Right: ");
        do {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        } while (temp != head);
        System.out.println("HEAD");
    }

    // Recursive helper
    static void traverseReverseUtil(Node current) {
        if (current.next != head)
            traverseReverseUtil(current.next);

        System.out.print(current.data + " -> ");
    }

    // Right to Left traversal
    static void traverseBackward() {
        if (head == null) {
            System.out.println("List is empty");
            return;
        }

        System.out.print("Right to Left: ");
        traverseReverseUtil(head);
        System.out.println("HEAD");
    }

    public static void main(String[] args) {
        traverseForward();
        traverseBackward();
    }
}

File: Circular_Linked_list/circular_singly_linked_list_traversal.c
// File Name: circular_singly_linked_list_traversal.c
// Program to traverse a circular singly linked list
// left to right and right to left (using recursion)

#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node *next;
};

struct node *head = NULL;

// Left to Right traversal
void traverse_forward() {
    if (head == NULL) {
        printf("List is empty\n");
        return;
    }

    struct node *temp = head;
    printf("Left to Right: ");
    do {
        printf("%d -> ", temp->data);
        temp = temp->next;
    } while (temp != head);
    printf("HEAD\n");
}

// Recursive helper for reverse traversal
void traverse_reverse_util(struct node *current, struct node *start) {
    if (current->next != start)
        traverse_reverse_util(current->next, start);

    printf("%d -> ", current->data);
}

// Right to Left traversal
void traverse_backward() {
    if (head == NULL) {
        printf("List is empty\n");
        return;
    }

    printf("Right to Left: ");
    traverse_reverse_util(head, head);
    printf("HEAD\n");
}

int main() {
    traverse_forward();
    traverse_backward();
    return 0;
}
